### PR TITLE
Tpetra: Reimplementation of MultiVector using WrappedDualView

### PR DESCRIPTION
**Pull request by csiefer2**
Issues: #9519 
Stakeholder Feedback: not yet!
Tests: Ran tests on OpenMP, Cuda (no-UVM) and Serial

@trilinos/tpetra Please read and give feedback!

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 1]
```diff
@@ -239,20 +275,106 @@ class WrappedDualView {
...
+      dualView.clear_sync_state();
+      dualView.modify_device();
+    }
     return dualView.view_device();
   }
 
+  template<class TargetDeviceType>
+  typename std::remove_reference<decltype(std::declval<DualViewType>().template view<TargetDeviceType>())>::type::const_type
+  getView (Access::ReadOnlyStruct s DEBUG_UVM_REMOVAL_ARGUMENT) const {    
+    if(std::is_same<TargetDeviceType,DeviceType>::value) {
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:289]: Unfortunately no.  There are a lot of calls to MultiVector::view<TargetDeviceType>() lying around.

File: packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp [hunk 2]
```diff
@@ -174,6 +174,8 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( MultiVector, HostView, LO, GO, Scalar , Node
   vector_t defaultVec(defaultMap);
   defaultVec.putScalar(scalar);
 
+  Kokkos::fence();
```
Comment (brian-kelley) [inline packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp:177]: Shouldn't getLocalViewHost below take care of this fence?

File: packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp [hunk 3]
```diff
@@ -242,14 +242,20 @@ namespace { // (anonymous)
 
       Teuchos::ArrayRCP<Teuchos::ArrayRCP<const Scalar> > viewsConst =
         X_noncontig->get2dView ();
-      TEST_ASSERT( ! X_noncontig->need_sync_host () );
+      if (! std::is_same<typename MV::dual_view_type::t_dev::memory_space,
```
Comment (brian-kelley) [inline packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp:245]: If the host/device memory spaces are the same, shouldn't need_sync_host always return false?

File: packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp [hunk 4]
```diff
@@ -242,14 +242,20 @@ namespace { // (anonymous)
...
+      }
     }
 
     out << "Test X_noncontig->get2dViewNonConst()" << endl;
     {
       Teuchos::ArrayRCP<Teuchos::ArrayRCP<Scalar> > viewsNonConst =
         X_noncontig->get2dViewNonConst ();
-      TEST_ASSERT( ! X_noncontig->need_sync_host () );
+      if (! std::is_same<typename MV::dual_view_type::t_dev::memory_space,
+	  typename MV::dual_view_type::t_host::memory_space>::value) {
```
Comment (brian-kelley) [inline packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp:256]: Same as above

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 5]
```diff
@@ -239,20 +304,124 @@ class WrappedDualView {
     if (iAmASubview()) {
       return getDeviceView(Access::ReadWrite);
     }
-    throwIfHostViewAlive();
-    dualView.clear_sync_state();
-    dualView.modify_device();
+    if(needsSyncPath()) {
+      if (deviceMemoryIsHostAccessible) Kokkos::fence();
+      throwIfHostViewAlive();
```
Comment (kddevin) [inline packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:309]: same as above.

**Review by kddevin — COMMENTED** [review]
Thanks for this work, @csiefer2 .  I added a few questions to those from @brian-kelley .

File: packages/nox/test/tpetra/ME_Tpetra_Heq_def.hpp [hunk 6]
```diff
@@ -276,7 +273,6 @@ evalModelImpl(const Thyra::ModelEvaluatorBase::InArgs<Scalar> &inArgs,
 
       // Reduce local contributions and fill the vector
       if (comm_->getSize() > 1) {
-        localSum->template modify<execution_space>();
         localSum->reduce();
         if (myRank == proc) {
           view_type source = localSum->template getLocalView<execution_space>();
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/nox/test/tpetra/ME_Tpetra_Heq_def.hpp]: Need to fix this!

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 7]
```diff
@@ -112,12 +119,20 @@ sync_device(DualViewType dualView) { }
...
 template <typename DualViewType>
 class WrappedDualView {
 public:
   using HostViewType = typename DualViewType::t_host;
   using DeviceViewType = typename DualViewType::t_dev;
 
+  using HostType   = typename HostViewType::device_type;
+  using DeviceType = typename DeviceViewType::device_type;
+
+  using DVT = DualViewType;
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:132]: Condense down to t_host and t_dev so things work more like DualView

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 8]
```diff
@@ -131,6 +146,13 @@ class WrappedDualView {
       dualView(originalDualView)
   { }
 
+  // Should this be expert only?
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:149]: Add expert-only comment

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 9]
```diff
@@ -196,10 +255,12 @@ class WrappedDualView {
     if (iAmASubview()) {
       return getHostView(Access::ReadWrite);
     }
-    throwIfDeviceViewAlive();
-    if (deviceMemoryIsHostAccessible) Kokkos::fence();
-    dualView.clear_sync_state();
-    dualView.modify_host();
+    if(needsSyncPath()) {
+      if (deviceMemoryIsHostAccessible) Kokkos::fence();      
+      throwIfDeviceViewAlive();
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:260]: We can switch back.

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 5 above)
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:309]: Ditto

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp [hunk 10]
```diff
@@ -317,15 +492,116 @@ class WrappedDualView {
...
+  ViewType getSubview2D(ViewType view, Kokkos::pair<int_type,int_type> offset0, Kokkos::pair<int_type,int_type> offset1) const {
+    return Kokkos::subview(view,offset0,offset1);
+  }
+
+
+  bool memoryIsAliased() const {
+    return deviceMemoryIsHostAccessible && dualView.h_view.data() == dualView.d_view.data();
+  }
 
+  bool needsSyncPath() const {
```
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:596]: Add explanation for what this actually means.  Not why we are checking things and how it effects performance.

File: packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp (hunk 2 above)
Comment (csiefer2) [reply packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp:177]: You are correct.  Will fix.

File: packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp (hunk 3 above)
Comment (csiefer2) [reply packages/tpetra/core/test/MultiVector/MultiVector_get2dView.cpp:245]: I'll take a second look at report back.

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:149]: done

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:260]: fixed

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 5 above)
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:309]: fixed

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:596]: done

File: packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp (hunk 2 above)
Comment (csiefer2) [reply packages/tpetra/core/test/MultiVector/MultiVector_LocalViewTests.cpp:177]: done

File: packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:132]: Done

**Review by kddevin — DISMISSED** [review]
My comments were addressed. Thanks, @csiefer2

**Comment by rppawlo** [comment]
@csiefer2 - this commit brought down empire testing last night. Errors are below. It looks like some of the deprecated code may have been removed from tptetra? I can probably just go in and change these on the empire side, but just wondering if this was intentional? 
```
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/em_solvers/ProjectionWrapper.cpp:1:
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/utils/Indexer.hpp:39:
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/utils/Typedefs.hpp:55:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVector.hpp:44:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVector_decl.hpp:47:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVectorSpace_decl.hpp:49:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Tpetra_MultiVector.hpp:1:
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Tpetra_MultiVector_decl.hpp:1532:36: error: no matching member function for call to 'modify'
00:52:18        view_.getDualView().template modify<TargetDeviceType> ();
00:52:18        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
00:52:18  /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/em_solvers/ProjectionWrapper.cpp:425:36: note: in instantiation of function template specialization 'Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::modify<Kokkos::Serial>' requested here
00:52:18          owned_solution_fields_[i]->modify<PHX::Device>();
00:52:18                                     ^
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Kokkos_DualView.hpp:780:8: note: candidate function template not viable: 'this' argument has type 'const Kokkos::DualView<double **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void>', but method is not marked const
00:52:18    void modify() {
00:52:18         ^
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/em_solvers/ProjectionWrapper.cpp:1:
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/utils/Indexer.hpp:39:
00:52:18  In file included from /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/utils/Typedefs.hpp:55:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVector.hpp:44:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVector_decl.hpp:47:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Thyra_TpetraVectorSpace_decl.hpp:49:
00:52:18  In file included from /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Tpetra_MultiVector.hpp:1:
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Tpetra_MultiVector_decl.hpp:1498:36: error: no matching member function for call to 'sync'
00:52:18        view_.getDualView().template sync<TargetDeviceType> ();
00:52:18        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
00:52:18  /scratch/empire-jenkins/empire-master-jobs/clang-opt-serial-static/src/EMPIRE/src/em_solvers/ProjectionWrapper.cpp:476:34: note: in instantiation of function template specialization 'Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::sync<Kokkos::Serial>' requested here
00:52:18        owned_solution_fields_[i]->sync<PHX::Device>();
00:52:18                                   ^
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Kokkos_DualView.hpp:606:8: note: candidate function template not viable: 'this' argument has type 'const Kokkos::DualView<double **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void>', but method is not marked const
00:52:18    void sync(const typename std::enable_if<
00:52:18         ^
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Kokkos_DualView.hpp:648:8: note: candidate template ignored: requirement '(!std::is_same<double **, double **>::value) || (std::is_same<Kokkos::Serial, int>::value)' was not satisfied [with Device = Kokkos::Serial]
00:52:18    void sync(const typename std::enable_if<
00:52:18         ^
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Kokkos_DualView.hpp:615:8: note: candidate function template not viable: requires at least argument 'exec', but no arguments were provided
00:52:18    void sync(const ExecutionSpace& exec,
00:52:18         ^
00:52:18  /projects/empire/installs/testing-potential-update/CLANG-10.0.0-RELEASE-SERIAL-STATIC/include/Kokkos_DualView.hpp:656:8: note: candidate function template not viable: requires at least argument 'exec', but no arguments were provided
00:52:18    void sync(const ExecutionSpace& exec,
00:52:18         ^
00:52:18  2 errors generated.
```

**Pull request merged**